### PR TITLE
Email queue client

### PR DESCRIPTION
**Pull request by kcherukupalli**
## What’s changing?

Emails can be customized based type of the request template. Hardcoded templates will be replaced in future sprint based on the requirements.
Still need to get the Mail client details to complete the Notifications.

## Why?

@nac-1116

## This PR has:

- [x] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [ ] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [x] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [x] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)

## Anything else?
Currently we are just logging the email content in Azure function.
Hardcoded test email templates and will customize in the future sprints based on the requirement.
All the documentation will be updated once we finish the Notifications.

### CODE REVIEW

File: iac/create-resources.bash [hunk 1]
```diff
@@ -208,9 +208,7 @@ main () {
   printenv PG_SECRET | tr -d '\n' | az keyvault secret set \
     --vault-name "$VAULT_NAME" \
     --name "$PG_SECRET_NAME" \
-    --file /dev/stdin \
-    --query id
-    #--value "$PG_SECRET"
+    --value "$PG_SECRET"
```
Comment (ryangeorgeharvey) [inline iac/create-resources.bash:211]: issue (blocking): revert these changes.

File: iac/iac-common.bash [hunk 2]
```diff
@@ -476,9 +487,7 @@ set_oidc_secret () {
   printf '%s' "$value" | az keyvault secret set \
     --vault-name "$VAULT_NAME" \
     --name "$secret_name" \
-    --file /dev/stdin \
-    --query id > /dev/null
-    #--value "$value"
+    --value "$value"
```
Comment (ryangeorgeharvey) [inline iac/iac-common.bash:490]: issue (blocking): revert these changes.

File: iac/iac-common.bash [hunk 3]
```diff
@@ -116,7 +121,7 @@ COLLAB_DB_NAME=collaboration
 LOG_ANALYTICS_WORKSPACE_NAME=${PREFIX}-log-analytics-workspace-${ENV}
 
 # Diagnostic Settings
-DIAGNOSTIC_SETTINGS_NAME="stream-diagnostic-logs"
+DIAGNOSTIC_SETTINGS_NAME="stream-logs-to-event-hub"
```
Comment (ryangeorgeharvey) [inline iac/iac-common.bash:124]: nitpick (non-blocking): Handling this in a different PR, so you can revert it.

File: iac/create-resources.bash [hunk 4]
```diff
@@ -359,6 +357,82 @@ main () {
   # Stream Event Grid topic diagnostic logs to Event Hub / Log Analytisc workspace
   update_event_grid_topic_diagnostic_settings "${event_grid_topic_id}"
 
+# Create Notifications API Function App
```
Comment (ryangeorgeharvey) [inline iac/create-resources.bash:360]: Suggestion: `function-notifications.json` looks good, but is essentially a copy of `function-states.json`, which in turn is just a copy of `function-match-res.json`. The only differences I see in the two files are the System Type Tag, which you added to IAC Common, and the Collaboration Database variable. 

Ideally, you could make minor changes to 'function-states.json' and name to 'function.json'. For the System Type Tag, you would pass the corresponding IAC Common variable, and for the configurations, you either pass the appropriate value or and empty string. Since 'function-match-res.json' could be supported with similar changes, if this work is not handled in this PR, we should log a JIRA story to track it.

File: iac/create-resources.bash [hunk 5]
```diff
@@ -359,6 +357,82 @@ main () {
...
+  
+ # Create a Subscription to the queue
+  az eventgrid event-subscription create \
+    --source-resource-id "${GRID_TOPIC_PROVIDERS}/Microsoft.EventGrid/topics/${CREATE_NOTIFICATIONS_TOPIC_NAME}" \
+    --name "$subn_name" \
+    --endpoint-type storagequeue \
+    --endpoint "$queueid" \
+  
+   # Configure log streaming for function app
+  event_grid_notification_system_topic_id=$(\
```
Comment (ryangeorgeharvey) [inline iac/create-resources.bash:421]: nitpick (non-blocking): Rename variable to `event_grid_notification_topic_id`. This is not a system topic.

File: iac/create-resources.bash [hunk 6]
```diff
@@ -359,6 +357,82 @@ main () {
...
+      coreResourceGroup="$RESOURCE_GROUP" \
+      diagnosticSettingName="${DIAGNOSTIC_SETTINGS_NAME}" \
+      eventHubAuthorizationRuleId="${EH_RULE_ID}" \
+      eventHubName="${EVENT_HUB_NAME}" \
+      sku="${STORAGE_SKU}" \
+      subnet="${FUNC_SUBNET_NAME}" \
+      vnet="${VNET_ID}" \
+      workspaceId="${LOG_ANALYTICS_WORKSPACE_ID}"
+ 
+  echo "Integrating ${NOTIFICATIONS_FUNC_APP_NAME} into virtual network"
```
Comment (ryangeorgeharvey) [inline iac/create-resources.bash:382]: issue (blocking): You created a Storage Account for the function in your ARM template. Here you are integrating that function into the VNET, which the Storage Account is also integrated into. However, you never remove public access from the Storage Account.

Reference https://github.com/18F/piipan/blob/33b85acacfb12a16dc5ffb2ba32870d6c6235a72/iac/create-resources.bash#L480 to fix.

File: iac/create-resources.bash (hunk 4 above)
Comment (ryangeorgeharvey) [inline iac/create-resources.bash:360]: nitpick (non-blocking): Some minor spacing issues with line starts (0 versus 1 versus 2).

**Review by ryangeorgeharvey — COMMENTED** [review]
nitpick (non-blocking): End of files should have newlines.

File: iac/create-resources.bash (hunk 4 above)
Comment (kcherukupalli) [reply iac/create-resources.bash]: Consolidated arm template files for Function Apps.

File: notifications/tests/Piipan.Notifications.Core.Integration.Tests/NotificationApiTes.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
+﻿using Azure.Messaging.EventGrid;
+using Microsoft.Extensions.Logging;
+using Moq;
+using Piipan.Notifications.Func.Api;
+using Xunit;
+
+namespace Piipan.Notifications.Core.IntegrationTests
+{
+    public class NotificationApiTes
```
Comment (bgfoshay) [inline notifications/tests/Piipan.Notifications.Core.Integration.Tests/NotificationApiTes.cs:9]: Nitpick: Looks like a typo in the class & file name

**Review by bgfoshay — APPROVED** [review]
Looks good. Thanks for all of the work that went into this change.

File: iac/arm-templates/function.json [hunk 8]
```diff
@@ -139,90 +148,95 @@
...
-            "properties": {
-                "serverFarmId": "[resourceId(parameters('coreResourceGroup'), 'Microsoft.Web/serverfarms', parameters('appServicePlanName'))]",
-                "httpsOnly": true,
-                "siteConfig": {
-                    "ftpsState": "Disabled",
-                    "alwaysOn": true,
-                    "minTlsVersion": "1.2",
-                    "appSettings": [
-                        /*
+      {
```
Comment (ryangeorgeharvey) [inline iac/arm-templates/function.json:151]: Nitpick (non-blocking): This is minor, and I know you copied some of this from another ARM file. However, there is a mix between 2 space and 4 space tabs in the file, and some minor bracket misalignments.  The issue happens only with the resource block from lines 151-239.

It's annoying but IaC bash files use 2 spaces for tabs, while the ARM file use 4 spaces.

**Review by ryangeorgeharvey — APPROVED** [review]
Thanks for taking the time to make the IaC changes and address my code review comments. I made one minor nitpick for your new ARM (function.json), but otherwise this looks great!
